llvm/lib/SYCLLowerIR/MutatePrintfAddrspace.cpp

```diff
@@ -0,0 +1,145 @@
+  auto *Literal = cast<Constant>(CI->getArgOperand(0));
+  Type *LiteralType = Literal->getType();
+  // Generate/find a correct literal
+  auto *CASLiteralType = PointerType::get(LiteralType->getPointerElementType(),
+                                          ConstantAddrspaceID);
+  StringRef CASLiteralName(Literal->getName().str() + "._AS2");
+  return CI->getModule()->getOrInsertGlobal(
+      CASLiteralName, CASLiteralType, [&] {
+        StringRef LiteralValue;
+        getConstantStringInfo(Literal, LiteralValue);
+        GlobalVariable *GV = Builder.CreateGlobalString(
+            LiteralValue, CASLiteralName, ConstantAddrspaceID, CI->getModule());
+        GV->setLinkage(GlobalValue::LinkageTypes::InternalLinkage);
+        GV->setUnnamedAddr(GlobalValue::UnnamedAddr::None);
+        return GV;
+      });
+}
+
+FunctionCallee getCASPrintfFunction(Function &GenericASPrintfFunc,
+                                    Type *CASLiteralTy) {
+  auto *CASPrintfFuncTy = FunctionType::get(GenericASPrintfFunc.getReturnType(),
+                                            CASLiteralTy, /*isVarArg=*/true);
+  FunctionCallee CASPrintfFunc =
+      GenericASPrintfFunc.getParent()->getOrInsertFunction(
+          "_Z18__spirv_ocl_printfPU3AS2Kcz", CASPrintfFuncTy,
+          GenericASPrintfFunc.getAttributes());
+  auto *Callee = cast<Function>(CASPrintfFunc.getCallee());
+  Callee->setCallingConv(CallingConv::SPIR_FUNC);
+  Callee->setDSOLocal(true);
+  return CASPrintfFunc;
+}
+
+CallInst *buildCASPrintfCall(IRBuilder<> &Builder, FunctionCallee CASPrintfFunc,
+                             Constant *CASLiteral, CallInst *CI) {
+  SmallVector<Value *, 4> CallOperands(CI->data_operands_begin(),
+                                       CI->data_operands_end());
+  CallOperands[0] = CASLiteral;
+  Builder.SetInsertPoint(CI);
+  CallInst *CASCall = Builder.CreateCall(CASPrintfFunc, CallOperands,
+                                         CI->getName().str() + "._AS2");
+  CASCall->setTailCall(true);
+  return CASCall;
+}
+} // namespace
+
+PreservedAnalyses
+SYCLMutatePrintfAddrspacePass::run(Module &M, ModuleAnalysisManager &MAM) {
+  size_t ReplacedCallsCount = 0;
+  SmallVector<Function *, 1> FunctionsToDrop;
+
+  for (Function &F : M) {
+    if (!F.isDeclaration())
+      continue;
+    if (!F.getName().contains("__spirv_ocl_printf"))
+      continue;
+    auto *LiteralType = F.getArg(0)->getType();
+    if (LiteralType->getPointerAddressSpace() == ConstantAddrspaceID)
+      // No need to replace the literal type and its printf users
+      continue;
+
+    IRBuilder<> Builder(M.getContext());
+    SmallVector<CallInst *, 8> CallInstsToDrop;
+    for (User *U : F.users()) {
+      if (!isa<CallInst>(U))
+        continue;
+      auto *CI = cast<CallInst>(U);
+      Constant *CASLiteral = getCASLiteral(Builder, CI);
+      FunctionCallee CASPrintfFunc =
+          getCASPrintfFunction(F, CASLiteral->getType());
```

Comment:
I suggest we move this definition out of the loop, because it is always the same on each iteration. Type of the first argument is known in advance and can be manually created instead of extracting from an existing call

Reply:
Done after investigation into the odd discrepancies between the pre-fetched `CASLiteralType` and the `getType()` results for the generated `CASLiteral`.